packages/tokamak/contracts-bedrock/src/L1/OptimismPortal.sol

```diff
@@ -383,9 +398,14 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
         bytes memory _data
     )
         public
-        payable
         metered(_gasLimit)
     {
+        // Lock token in this contract
+        if (_value > 0)
+        {
+            IERC20(tonAddress).safeTransferFrom(msg.sender, address(this), _value);
+        }
```

Comment:
I think we should move this function after some `require` function above to prevent if this transaction isn't valid, we don't transfer from sender to PORTAL

Reply:
Thank you!
Could you explain more about how to use require function here?

Reply:
@nguyenzung I mean in this function, there are some conditions, so I think we should move this code after that
https://github.com/tokamak-network/tokamak-titan-canyon/blob/b325fbce26880ec11e30ad6563a61851dee9fe38/packages/tokamak/contracts-bedrock/src/L1/OptimismPortal.sol#L411

😊

Reply:
Oh, I understood your idea,
I think we don't need to do that because if there is any thing that not satisfied conditions, the whole transaction will be reverted!

Reply:
I think we can use this code,,,
Is it reverted entire state changes when it fail?

Reply:
Oh I got it, so many thanks